feat: add tray and menu javascript APIs and events, closes #6617

amrbashir opened the pull request:
closes #6617

TODOs:
- [x] JS docs
 
<!--
Update "[ ]" to "[x]" to check a box

Please make sure to read the Pull Request Guidelines: https://github.com/tauri-apps/tauri/blob/dev/.github/CONTRIBUTING.md#pull-request-guidelines
-->

### What kind of change does this PR introduce?
<!-- Check at least one. If you are introducing a new binding, you must reference an issue where this binding has been proposed, discussed and approved by the maintainers. -->

- [ ] Bugfix
- [ ] Feature
- [ ] Docs
- [ ] New Binding issue #___
- [ ] Code style update
- [ ] Refactor
- [ ] Build-related changes
- [ ] Other, please describe:

### Does this PR introduce a breaking change?
<!-- If yes, please describe the impact and migration path for existing applications in an attached issue. -->

- [ ] Yes, and the changes were approved in issue #___
- [x] No

### Checklist
- [ ] When resolving issues, they are referenced in the PR's title (e.g `fix: remove a typo, closes #___, #___`)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).
- [ ] I have added a convincing reason for adding this feature, if necessary

### Other information

amrbashir committed:
feat: add tray and menu javascript APIs and events

closes #6617

amrbashir committed:
some progress on menu and submenu

amrbashir committed:
implement the rest, needs testing

amrbashir committed:
implement js menu events

amrbashir committed:
update example api to latest vite-svelte template

amrbashir committed:
fix ipc protocol for empty args

amrbashir committed:
add anyhow error variant

amrbashir committed:
fix menu js implementation

amrbashir committed:
remove console.log

amrbashir committed:
lint and clippy

amrbashir committed:
docs

amrbashir committed:
fix cargo.toml

amrbashir committed:
generated files

amrbashir committed:
fix check license header script

amrbashir committed:
fix android and ios build

amrbashir commented:
Now that the implementation is done without depending on the `window` plugin, I still believe the `window` plugin (plus some others) should be added back to core, also notice that the `tray` plugin is dependent on the `menu` plugin so both will need to be in coupled together.

Also the `menu` plugin ideally should use the `Window` and `Position` types from the `window` plugin and the `window` plugin should depend on the `menu` plugin so that we can provide a window config option to use a specific menu when creating the window.

All this could be solved by just adding them all into one single package in core. These are essential to any tauri app and so should be provided by default in core.

amrbashir committed:
change file

amrbashir committed:
change tag

amrbashir committed:
Merge branch 'dev' into feat/tray-menu-js-apis

amrbashir committed:
fmt

lucasfernog commented on core/tauri-macros/src/lib.rs:
```diff
@@ -89,3 +90,64 @@ pub fn default_runtime(attributes: TokenStream, input: TokenStream) -> TokenStre
   runtime::default_runtime(attributes, input).into()
 }
+
+/// Accepets a closure-like syntax to call arbitray code on a menu item
```
```suggestion
/// Accepts a closure-like syntax to call arbitrary code on a menu item
```

lucasfernog commented on core/tauri-macros/src/lib.rs:
```diff
@@ -89,3 +90,64 @@ pub fn default_runtime(attributes: TokenStream, input: TokenStream) -> TokenStre
 }
+
+/// Accepets a closure-like syntax to call arbitray code on a menu item
+/// after mathcing against `kind` and retreiveing it from `resources_table` using `rid`.
```
```suggestion
/// after matching against `kind` and retrieving it from `resources_table` using `rid`.
```

lucasfernog commented on core/tauri-macros/src/lib.rs:
```diff
@@ -89,3 +90,64 @@ pub fn default_runtime(attributes: TokenStream, input: TokenStream) -> TokenStre
+/// Accepets a closure-like syntax to call arbitray code on a menu item
+/// after mathcing against `kind` and retreiveing it from `resources_table` using `rid`.
+///
+/// You can optionally pass a third paraemeter to select which item kinds
```
```suggestion
/// You can optionally pass a third parameter to select which item kinds
```

lucasfernog commented on core/tauri-macros/src/lib.rs:
```diff
@@ -89,3 +90,64 @@ pub fn default_runtime(attributes: TokenStream, input: TokenStream) -> TokenStre
+/// after mathcing against `kind` and retreiveing it from `resources_table` using `rid`.
+///
+/// You can optionally pass a third paraemeter to select which item kinds
+/// to match agains, by providing a `|` separated list of item kinds
```
```suggestion
/// to match against, by providing a `|` separated list of item kinds
```

lucasfernog commented on core/tauri-macros/src/menu.rs:
```diff
@@ -0,0 +1,104 @@
+    kinds.extend(defaults.clone());
+  }
+
+  let mut has_negated: bool = false;
+  for NegatedIdent(negated, _) in &kinds {
+    if *negated && !has_negated {
+      has_negated = true;
+    }
+  }
```
```suggestion
  let has_negated = kinds.iter().any(|n| n.0);
```

amrbashir replied:
I srsly don't know what I was thinking writing that code, just kill me already 😅

lucasfernog commented on core/tauri-macros/src/menu.rs:
```diff
@@ -0,0 +1,104 @@
+      i if i == "Check" && !nident.0 => Some((i, Ident::new("CheckMenuItem", Span::call_site()))),
+      i if i == "Icon" && !nident.0 => Some((i, Ident::new("IconMenuItem", Span::call_site()))),
+      _ => None,
+    })
```
```suggestion
    .filter_map(|nident| {
      if nident.0 {
        None
      } else {
        match nident.1 {
          i if i == "MenuItem" => Some((i, Ident::new("MenuItem", Span::call_site()))),
          i if i == "Submenu" => Some((i, Ident::new("Submenu", Span::call_site()))),
          i if i == "Predefined" => Some((i, Ident::new("PredefinedMenuItem", Span::call_site()))),
          i if i == "Check" => Some((i, Ident::new("CheckMenuItem", Span::call_site()))),
          i if i == "Icon" => Some((i, Ident::new("IconMenuItem", Span::call_site()))),
          _ => None,
        }
      }
    })
```

lucasfernog commented on core/tauri-macros/src/menu.rs:
```diff
@@ -0,0 +1,104 @@
+    .unzip();
+
+  quote! {
+    match kind {
```
I don't like how it assumes `kind`, `rid` and `resources_table` exists instead of taking them as arguments.. but this is internal so it's ok.

amrbashir replied:
yeah I don't like it either, maybe I need to spend some time thinking about improving it without adding much noise in the call site

lucasfernog commented on core/tauri/src/resources/mod.rs:
```diff
@@ -0,0 +1,159 @@
+
+/// Resources are Rust objects that are stored in [ResourceTable] and managed by tauri.
+/// They are identified in JS by a numeric ID (the resource ID, or rid).
+/// Resources can be created in endpoints. Resources can also be retrieved in endpoints by
```
```suggestion
/// Resources can be created in commands. Resources can also be retrieved in commands by
```

lucasfernog commented on core/tauri/src/resources/mod.rs:
```diff
@@ -0,0 +1,159 @@
+  pub(crate) fn downcast_arc<'a, T: Resource>(self: &'a Arc<Self>) -> Option<&'a Arc<T>> {
+    if self.is::<T>() {
+      let ptr = self as *const Arc<_> as *const Arc<T>;
+      // TODO(piscisaureus): safety comment
```
let's resolve this :)

lucasfernog commented on core/tauri/src/resources/mod.rs:
```diff
@@ -0,0 +1,159 @@
+  /// The resource type is erased at runtime and must be statically known
+  /// when retrieving it through `get()`.
+  ///
+  /// Returns a unique resource ID, which acts as a key for this resource.
```
```suggestion
  /// Returns an unique resource ID, which acts as a key for this resource.
```

lucasfernog commented on core/tauri/src/resources/mod.rs:
```diff
@@ -0,0 +1,159 @@
+
+  /// Returns a reference counted pointer to the resource of type `T` with the
+  /// given `rid`. If `rid` is not present or has a type different than `T`,
+  /// this function returns `None`.
```
```suggestion
  /// this function returns [`Error::BadResourceId`].
```

lucasfernog commented on core/tauri/src/resources/mod.rs:
```diff
@@ -0,0 +1,159 @@
+  // /// assume that `Arc::strong_count(&returned_arc)` is always equal to 1 on success.
+  // /// In particular, be really careful when you want to extract the inner value of
+  // /// type `T` from `Arc<T>`.
+  // pub(crate) fn take<T: Resource>(&mut self, rid: ResourceId) -> Result<Arc<T>, Error> {
```
should we delete this?

amrbashir replied:
I left it there in case we ever need but I guess we can re-implement it

lucasfernog commented on core/tauri/src/resources/plugin.rs:
```diff
@@ -0,0 +1,22 @@
+use super::ResourceId;
+
+#[command(root = "crate")]
+fn close<R: Runtime>(app: AppHandle<R>, rid: ResourceId) -> crate::Result<()> {
```
this is problematic, an attacker can close all resources by iterating on IDs until an error is returned.

maybe we should make the ID random instead?

amrbashir replied:
Closing resource is not that bad right?

Also for reference, deno is using the same ID strategy and though they also provide `.close()` you can't really invoke the underlying mechanism of it (compared to `invoke("plugin:resources|close"`). 

So in order to simplify the implementation and keep using incrementing ID numbers, we can hide the invoke endpoint and randomize it on each launch of the app.

lucasfernog commented on core/tauri/src/menu/plugin.rs:
```diff
@@ -0,0 +1,637 @@
+// Copyright 2019-2023 Tauri Programme within The Commons Conservancy
```
this file is kinda big, can we move the commands to their own module?

amrbashir replied:
the commands make most of the file, so moving them into another module will still be a big file

lucasfernog commented on core/tauri/src/tray/plugin.rs:
```diff
@@ -0,0 +1,202 @@
+  };
+
+  builder = builder.on_tray_event(|tray, e| {
+    let _ = tray.app_handle().emit_all("tauri://tray", e);
```
I think we should use a channel here instead, so we know only the tray icon caller have access to these events

also simplifies the JS APi a lot :)

lucasfernog commented on core/tauri/src/menu/plugin.rs:
```diff
@@ -0,0 +1,637 @@
+  Builder::new("menu")
+    .on_event(|app, e| {
+      if let RunEvent::MenuEvent(e) = e {
+        let _ = app.emit_all("tauri://menu", e);
```
consider using a channel instead (maybe have a dedicated list of channels in the resources table)

amrbashir committed:
Apply suggestions from code review

Co-authored-by: Lucas Fernandes Nogueira <lucas@tauri.app>

amrbashir committed:
Update core/tauri/src/resources/mod.rs

Co-authored-by: Lucas Fernandes Nogueira <lucas@tauri.app>

amrbashir committed:
Update core/tauri-macros/src/menu.rs

Co-authored-by: Lucas Fernandes Nogueira <lucas@tauri.app>

amrbashir committed:
Update core/tauri/src/resources/mod.rs

Co-authored-by: Lucas Fernandes Nogueira <lucas@tauri.app>

amrbashir committed:
Update core/tauri/src/resources/mod.rs

Co-authored-by: Lucas Fernandes Nogueira <lucas@tauri.app>

amrbashir committed:
Update core/tauri-macros/src/menu.rs

Co-authored-by: Lucas Fernandes Nogueira <lucas@tauri.app>

amrbashir committed:
lints, safety comments, remove unsused code

amrbashir commented:
I think this PR needs to wait until the refactors PR are merged in.

amrbashir committed:
whitespace and removed clippy lint

lucasfernog committed:
Merge remote-tracking branch 'origin/dev' into feat/tray-menu-js-apis

lucasfernog committed:
typo

lucasfernog committed:
use channels

lucasfernog committed:
lint

lucasfernog committed:
private constructors

lucasfernog committed:
basic tray example

FabianLars commented on core/tauri/src/resources/mod.rs:
```diff
@@ -0,0 +1,143 @@
+  /// The resource type is erased at runtime and must be statically known
+  /// when retrieving it through `get()`.
+  ///
+  /// Returns an unique resource ID, which acts as a key for this resource.
```
idk why the github notification showed me a single change (`a` to `an`) and literally nothing else, but that's the only reason i noticed it i guess. `a unique` was correct because it starts with a `yu` sound.
And people say german is complicated... 😡 
```suggestion
  /// Returns a unique resource ID, which acts as a key for this resource.
```

lucasfernog committed:
Update core/tauri/src/resources/mod.rs [skip ci]

Co-authored-by: Fabian-Lars <fabianlars@fabianlars.de>

lucasfernog committed:
basic menu example

lucasfernog committed:
process tray icon

lucasfernog committed:
fix menu events

lucasfernog committed:
resolve TODOs, add popup example [skip ci]

lucasfernog committed:
Merge remote-tracking branch 'origin/dev' into feat/tray-menu-js-apis

[skip ci]

lucasfernog committed:
camelcase [skip ci]

lucasfernog committed:
popup freezes on macos :( gotta be async

lucasfernog committed:
fix menu example on macos

lucasfernog committed:
Merge remote-tracking branch 'origin/dev' into feat/tray-menu-js-apis

lucasfernog committed:
fix popup on macos

lucasfernog committed:
lint [skip ci]

lucasfernog committed:
Merge remote-tracking branch 'origin/dev' into feat/tray-menu-js-apis

amrbashir committed:
chore: cleanup

amrbashir committed:
cleanup menu JS APIs

amrbashir committed:
dedup package and run ts:check in CI

amrbashir committed:
Merge branch 'dev' into feat/tray-menu-js-apis

amrbashir committed:
lciense headers and fix lint

amrbashir committed:
generate bundle

amrbashir committed:
use explicit idents

amrbashir committed:
fix api example port

amrbashir committed:
fix circular modules dependency

amrbashir committed:
update api example

amrbashir committed:
licesne and lint

amrbashir committed:
fix build

lucasfernog committed:
update api dist [skip ci]

lucasfernog committed:
fix predefined

lucasfernog committed:
allow using objects directly for menu item construction for static menus

allows something like:
```javascript
await Menu.new({
  items: [
    { item: 'Copy' },
    { text: 'Something', action: () => console.log('something clicked') },
    {
      id: 'x',
      icon: '/home/lucas/projects/tauri/tauri/examples/.icons/32x32.png',
      text: 'Tauris',
      action: () => console.log('tauri clicked')
    },
    {
      checked: true,
      text: 'chec',
      action: () => console.log('chec clicked')
    },
    {
      text: 'Sub',
      items: [
        {
          item: 'Paste'
        },
        { text: 'subsubsub', action: () => console.log('sub clicked') }
      ]
    }
  ]
})
```

lucasfernog commented:
We're missing a lot of documentation here, but I wanna merge it anyway because we need to do a complete pass on API examples.

I like how this can be dynamic with the MenuItem classes (so you can update, share items), or static with raw objects after https://github.com/tauri-apps/tauri/pull/7709/commits/d0eeaba4198c3b04b5f252bd87f5ff87b3359fe2

lucasfernog committed:
fix mobile channel load

lucasfernog committed:
avoid panics

lucasfernog committed:
inline MenuItemKind

lucasfernog committed:
cleanup new channel interface

lucasfernog committed:
fix pre-commit hook

lucasfernog committed:
camel case

lucasfernog committed:
merge set_icon and set_native_icon

lucasfernog committed:
fix items(), get()

lucasfernog committed:
fix example [skip ci]
